test/system/src/test/java/io/pravega/test/system/ConsumptionBasedRetentionWithMultipleReaderGroupsTest.java

```diff
@@ -0,0 +1,287 @@
+            assertEquals("data of size 30", read.getEvent());
+        }
+        log.info("{} generating 1st stream-cuts for {}/{}", READER_GROUP_1, SCOPE, STREAM);
+        CompletableFuture<Map<Stream, StreamCut>> futureCuts1 = readerGroup1.generateStreamCuts(streamCutExecutor);
+        // Wait for 5 seconds to force reader group state update. This will allow for the silent
+        // checkpoint event generated as part of generateStreamCuts to be picked and processed.
+        Exceptions.handleInterrupted(() -> TimeUnit.SECONDS.sleep(5));
+        read = reader1.readNextEvent(READ_TIMEOUT);
+        assertEquals("data of size 30", read.getEvent());
+        assertTrue("Stream-cut generation did not complete for reader group 1", Futures.await(futureCuts1, 10000));
+        Map<Stream, StreamCut> streamCuts1 = futureCuts1.join();
+        log.info("{} generated 1st Stream cut at -> {}", READER_GROUP_1, streamCuts1);
+
+        // Read four events with reader2.
+        for (int i = 1; i <= 4; i++) {
+            read2 = reader2.readNextEvent(READ_TIMEOUT);
+            assertEquals("data of size 30", read2.getEvent());
+        }
+
+        log.info("{} generating 1st stream-cuts for {}/{}", READER_GROUP_2, SCOPE, STREAM);
+        CompletableFuture<Map<Stream, StreamCut>> futureCuts2 = readerGroup2.generateStreamCuts(streamCutExecutor);
+        // Wait for 5 seconds to force reader group state update. This will allow for the silent
+        // checkpoint event generated as part of generateStreamCuts to be picked and processed.
+        Exceptions.handleInterrupted(() -> TimeUnit.SECONDS.sleep(5));
+        read2 = reader2.readNextEvent(READ_TIMEOUT);
+        assertEquals("data of size 30", read2.getEvent());
+        assertTrue("Stream-cut generation did not complete for reader group 2", Futures.await(futureCuts2, 10000));
+        Map<Stream, StreamCut> streamCuts2 = futureCuts2.join();
+        log.info("{} updating its retention stream-cut to {}", READER_GROUP_1, streamCuts1);
+        readerGroup1.updateRetentionStreamCut(streamCuts1);
+        log.info("{} updating its retention stream-cut to {}", READER_GROUP_2, streamCuts2);
+        readerGroup2.updateRetentionStreamCut(streamCuts2);
+
+        // Retention set has one stream cut at 0/210
+        // READER_GROUP_1 updated stream cut at 0/90, READER_GROUP_2 updated stream cut at 0/120
+        // Subscriber lower bound is 0/90, truncation should happen at this point
+        // The timeout is set to 3 minutes a little longer than the retention period which is set to 2 minutes
+        // in order to confirm that the retention has taken place.
+        // Check to make sure truncation happened at streamcut generated by first subscriber
+        AssertExtensions.assertEventuallyEquals("Truncation did not take place at offset 90.", true, () -> controller.getSegmentsAtTime(
+                        new StreamImpl(SCOPE, STREAM), 0L).join().values().stream().anyMatch(off -> off >= 90),
```

Comment:
Please update this condition `off >= 90`. This may succeed if there is no truncation happens.

Reply:
updated it to `anyMatch(off -> off == 90)`.

Reply:
controller.getSegmentsAtTime will return the segment and its initial offset. So if truncation happens then for segment 0 offset should be 90.